src/library/scala/collection/Iterable.scala

```diff
@@ -807,6 +807,16 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     */
   def inits: Iterator[C] = iterateUntilEmpty(_.init)
 
+  /** Applies a side-effecting function to each element in this collection. Strict
+    * collections like Seq's will apply `f` to their elements immediately, while
+    * lazy collections like Views will only apply `f` when elements are evaluated.
+    *
+    * @param f the function to apply to each element
+    * @tparam U the return type of f
+    * @return The same logical collection as this
+    */
+  def tapEach[U](f: A => U): C = fromSpecific(new View.TapEach(this, f))
```

Comment:
you could potentially optimise this for strict collections by doing

```scala
  def tapEach[U](f: A => U): C = knownSize match {
    case 0          => this
    case n if n < 0 => fromSpecific(new View.TapEach(this, f))
    case n if n > 0 =>
      foreach(f)
      this
```

Reply:
it's not perfect (e.g. it misses `List`), but it might be better than nothing?

Reply:
I think that this size optimization could result in surprising behaviour if the underlying collection is mutated but lazy. In that case, when the empty collection has elements inserted, and then the lazy collection forces, `f` will not be called. I do not know if there's actually any collections like this however...